src/prefect/serialization/run_config.py

```diff
@@ -0,0 +1,63 @@
+from typing import Callable
+
+from marshmallow import fields, Schema, pre_dump, post_load, ValidationError
+
+from prefect.utilities.serialization import JSONCompatible
+from prefect.run_configs import RunConfig, KubernetesJob
+
+
+class RunConfigSchema(Schema):
+    cls_to_spec_schema = {}
+    kind_to_cls = {}
+
+    labels = fields.List(fields.String())
+    kind = fields.String()
+    spec = fields.Dict(keys=fields.String())
+
+    @pre_dump
+    def pre_dump(self, obj: RunConfig, **kwargs) -> dict:
+        cls = type(obj)
+        if cls in self.cls_to_spec_schema:
+            schema = self.cls_to_spec_schema[cls]()
+            spec = schema.dump(obj)
+            return {
+                "labels": getattr(obj, "labels", None),
+                "kind": cls.__name__,
+                "spec": spec,
+            }
+        raise ValidationError(f"Don't know how to serialize {cls.__name__} objects")
+
+    @post_load
+    def post_load(self, obj: dict, **kwargs) -> RunConfig:
+        kind = obj.get("kind")
+        if kind is None:
+            raise ValidationError("No `RunConfig` kind found")
+        elif kind not in self.kind_to_cls:
+            raise ValidationError(f"Unknown `RunConfig` kind '{kind}'")
+        cls = self.kind_to_cls[kind]
+        labels = obj.get("labels")
+        spec = obj.get("spec", {})
+        return cls(labels=labels, **spec)
+
+
+def register(run_config_cls: Callable) -> Callable:
```

Comment:
Hm I think a lot of this file might be recreating the logic for the `OneofSchema` object; see, e.g., this example: https://github.com/PrefectHQ/prefect/blob/master/src/prefect/serialization/state.py#L183

Reply:
From my comment above:

> Serializers have been updated to support serializing this new field along with the flow. Note that these look a bit different than the existing marshmallow serializers - I wanted the output json to match a certain schema, and the existing marshmallow machinery didn't support that as easily as one would hope (too much magic for my liking). In the future it should be straightforward to port this to pydantic.

TL;DR - I think designing a json schema is something that should be thought about as an api decision, rather than relying on how marshmallow conventions to decide the schema for us. I don't particularly care for`OneOfSchema` mixes the type switch with all the type-specific-fields (if we want to add more cross-run-config fields, we run the risk of conflicting with backend-specific args). This separates them so that things that belong on all `RunConfig` objects are at the top-level (and nothing else), and everything backend specific is in the `spec` field. Separation of concerns. The code here is fairly short and self contained, and upon switching to pydantic (which I think we're planning on doing?) a lot of these complications will just go away.

Reply:
After thinking a bit more, I think consistency with other objects may be a stronger argument than whatever complaints I have against `OneOfSchema`. Could go either way here.